src/projectworks.ts

```diff
@@ -0,0 +1,64 @@
+import fetch, { Headers, Response } from 'node-fetch';
+import { URLSearchParams } from 'url';
+import { hasIn, isArray, isObject } from './lowerDash';
+import type { ProjectWorksLeave, ProjectWorksUser } from './types';
+
+const PW_URL = 'https://api.projectworksapp.com';
+const PW_USERNAME = process.env.PROJECTWORKS_USERNAME;
+const PW_PASSWORD = process.env.PROJECTWORKS_PASSWORD;
+
+const apiGet = async (path: string): Promise<Response> => {
+  const headers = new Headers();
+
+  headers.append('Accept', 'application/json');
+  headers.append(
+    'Authorization',
+    `Basic ${Buffer.from(
+      `${PW_USERNAME as string}:${PW_PASSWORD as string}`
+    ).toString('base64')}`
+  );
+
+  return fetch(`${PW_URL}${path}`, { headers });
+};
+
+const isValidUser = (maybeUser: unknown): maybeUser is ProjectWorksUser => {
+  return isObject(maybeUser) && hasIn(maybeUser, 'UserID');
+};
+
+const isValidLeaves = (
+  maybeLeaves: unknown
+): maybeLeaves is ProjectWorksLeave[] => {
+  return (
+    isArray(maybeLeaves) &&
+    isObject(maybeLeaves[0]) &&
+    hasIn(maybeLeaves[0], 'LeaveID')
+  );
+};
+
+export const fetchUser = async (userId: number): Promise<ProjectWorksUser> => {
+  const rawUser = await apiGet(`/api/v1.0/Users/${userId}`);
+
+  const user: unknown = await rawUser.json();
+
+  if (isValidUser(user)) {
+    return user;
+  }
+
+  throw new Error('Failed to get User');
```

Comment:
I know we've just gone back and forward refactoring these validation checks, but I'm curious to why you're actually doing them in the first place vs just defining an interface with the appropriate properties? I'd be pretty concerned if the public PW API is so unstable...

Reply:
My goal was "Take the JS that already works and just make it TS with as few changes as possible" so I've ended up here through a number of rounds of 1) make TS happy, 2) refactor, 3) goto 1.

That being said, I have time now and I'm keen to learn a bit. I should check the response for HTTP 400/500 (which I will do) which leaves me satisfying the types. I'm not sure I follow your question about defining an interface with appropriate properties. 

My understanding is that I need to get from a `Promise<any>` (as returned by `.json()` to a `ProjectWorksUser` (which I have already defined). 

From my reading of the TS docs I can either just use a [type assertion](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions) to do this e.g.

```ts
const user = (await response.json()) as ProjectWorksUser; // json() returns Promise<any>
```

or use [type narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html). Narrowing seemed more robust so I followed that path. That being said, re-reading those docs today, I see that type narrowing is really more excluding types from a given union type rather than "this blob of JSON isn't what I expected".

Is this a place to just use the type assertion?

Reply:
Yup, what you've said pretty much hits the nail on the head - validating the response isn't strictly wrong, but one of the implications of a publicly documented API is a level of stability so that we can forgo the extra complexity of that validation safely;  so casting is the way to go here imo